### PR TITLE
[admission-policy-engine] add policy for replicas limits

### PR DESCRIPTION
**Pull request by sh1kel**
## Description
New operation policy for checking replica values for Deployments, ReplicaSets and StatefulSets

## Why do we need it, and what problem does it solve?
We would need to control low and high ranges of replicas to prevent zero replicas or check too high values.

## What is the expected result?
If replica value is not in range we'll get an error

## Checklist
- [x] The code is covered by unit tests.
- [ ] e2e tests passed.
- [x] Documentation updated according to the changes.
- [x] Changes were tested in the Kubernetes cluster manually.

## Changelog entries
New operation policy for replica value checking
```yaml
apiVersion: deckhouse.io/v1alpha1
kind: OperationPolicy
metadata:
  name: foo
spec:
  policies:
    replicaLimits:
      minReplicas: 1
      maxReplicas: 3
  match:
    namespaceSelector:
      matchNames: ["default"]
```

```sh
$ kubectl -n default patch deployment echo-server  --type='merge' -p '{"spec":{"replicas":5}}'
Error from server (Forbidden): admission webhook "admission-policy-engine.deckhouse.io" denied the request: [foo] The provided number of replicas is not allowed for Deployment: echo-server. Allowed ranges: {"ranges": [{"maxReplicas": 3, "minReplicas": 1}]}
```

```yaml
apiVersion: deckhouse.io/v1alpha1
kind: OperationPolicy
metadata:
  name: foo
spec:
  policies:
    replicaLimits:
      minReplicas: 3
  match:
    namespaceSelector:
      matchNames: ["default"]
```

```sh
$ kubectl -n default patch deployment echo-server  --type='merge' -p '{"spec":{"replicas":5}}'
deployment.apps/echo-server patched

$ kubectl -n default patch deployment echo-server  --type='merge' -p '{"spec":{"replicas":1}}'
Error from server (Forbidden): admission webhook "admission-policy-engine.deckhouse.io" denied the request: [foo] The provided number of replicas is not allowed for Deployment: echo-server. Allowed ranges: {"ranges": [{"minReplicas": 3}]}
```

```changes
section:admission-policy-engine
type:  feature
summary: operation policy for replica value checking
```

### CODE REVIEW

File: modules/015-admission-policy-engine/crds/operation-policy.yaml [hunk 1]
```diff
@@ -171,6 +171,19 @@ spec:
                     checkContainerDuplicates:
                       type: boolean
                       description: "Check container names and env variables for duplicates."
+                    replicaLimits:
+                      type: object
+                      description: "Allowed ranges for numbers of replicas.  Values are inclusive."
+                      items:
```
Comment (yalosev) [inline modules/015-admission-policy-engine/crds/operation-policy.yaml:177]: Lets make it not an array

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @yalosev. I've started the workflow for label 'deploy/web/test'...

File: modules/015-admission-policy-engine/crds/operation-policy.yaml [hunk 2]
```diff
@@ -171,6 +171,16 @@ spec:
                     checkContainerDuplicates:
                       type: boolean
                       description: "Check container names and env variables for duplicates."
+                    replicaLimits:
+                      type: object
+                      description: "A range of allowed replicas.  Values are inclusive."
+                      properties:
+                        min_replicas:
```
Comment (yalosev) [inline modules/015-admission-policy-engine/crds/operation-policy.yaml:178]: camel case

File: modules/015-admission-policy-engine/crds/operation-policy.yaml [hunk 3]
```diff
@@ -171,6 +171,16 @@ spec:
...
+                          description: "The minimum number of replicas allowed, inclusive."
```
Comment (yalosev) [inline modules/015-admission-policy-engine/crds/operation-policy.yaml:179]: and ru docs required

**Review by yalosev — APPROVED** [review]

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @yalosev. I've started the workflow for label 'e2e/run/aws'...

File: modules/015-admission-policy-engine/crds/operation-policy.yaml [hunk 4]
```diff
@@ -171,6 +171,16 @@ spec:
                     checkContainerDuplicates:
                       type: boolean
                       description: "Check container names and env variables for duplicates."
+                    replicaLimits:
```
Comment (nabokihms) [inline modules/015-admission-policy-engine/crds/operation-policy.yaml:174]: Why is the `Limits` in the plural form? Previously, there were many limits, but now it seems like there is only one. Or is it about upper limit and lower limit?

**Review by nabokihms — APPROVED** [review]

File: modules/015-admission-policy-engine/crds/operation-policy.yaml (hunk 4 above)
Comment (sh1kel) [reply modules/015-admission-policy-engine/crds/operation-policy.yaml:174]: Exactly, there are two limits for one **Replicas** parameter

**Pull request merged**